karapace/kafka/producer.py

```diff
@@ -59,3 +62,69 @@ def partitions_for(self, topic: str) -> dict[int, PartitionMetadata]:
             return self.list_topics(topic).topics[topic].partitions
         except KafkaException as exc:
             raise_from_kafkaexception(exc)
+
+
+class AsyncKafkaProducer:
+    """An async wrapper around `KafkaProducer` built on top of confluent-kafka.
+
+    Calling `start` on an `AsyncKafkaProducer` instantiates a `KafkaProducer`
+    and starts a poll-thread.
+
+    The poll-thread continuously polls the underlying producer so buffered messages
+    are sent and asyncio futures returned by the `send` method can be awaited.
+    """
+
+    def __init__(
+        self,
+        bootstrap_servers: Iterable[str] | str,
+        loop: asyncio.AbstractEventLoop | None = None,
+        **params: Unpack[KafkaClientParams],
+    ) -> None:
+        self.loop = loop or asyncio.get_running_loop()
+
+        self.stopped = Event()
+        self.poll_thread = Thread(target=self.poll_loop)
+
+        self.producer: KafkaProducer | None = None
+        self._bootstrap_servers = bootstrap_servers
+        self._producer_params = params
+
+    def _start(self) -> None:
+        self.producer = KafkaProducer(self._bootstrap_servers, **self._producer_params)
+        self.stopped.clear()
```

Comment:
nit (non-blocking): Should we really allow restarting rather than erroring out, it seems some resource management will get complex if we allow restarting, and I believe the subsequent call to `self.poll_thread.start()` will currently fail because it tries restart an existing thread? Not sure we need to make something explicit, but this `.clear()` call could potentially be removed, and for safe usage it might make sense to raise if the event is set when entering `_start()`?

Reply:
Good point about the thread restart, adding a guard here.